### ISSUE TITLE
Unrelated element gets nan gradient

### ISSUE THREAD

**Issue by Steve-Tod** [pytorch/pytorch#26799]
## 🐛 Bug

 I divide an element by zero and get an inf. But this inf is not used in the following calculation. However, after backward, I still get a nan gradient for this element.

## To Reproduce

```python
import torch
divide_tensor = torch.Tensor([1.0, 0.0, 1.0])
weight = torch.ones(3)
weight.requires_grad = True
tmp = weight / divide_tensor
loss = tmp[0]
loss.backward()
print(weight.grad)
```

## Expected behavior

I expect gradient [1., 0., 0.] because second and third elements of weight are not related to loss.

## Environment

PyTorch version: 1.0.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration:
GPU 0: TITAN Xp
GPU 1: TITAN Xp
GPU 2: TITAN Xp
GPU 3: TITAN Xp

Nvidia driver version: 410.78
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] torch==1.0.1
[pip] torchvision==0.2.1
[conda] _tflow_select             2.3.0                       mkl
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.0.2            py36h7b6447c_0
[conda] mkl_fft                   1.0.12           py36ha843d7b_0
[conda] mkl_random                1.0.2            py36hd81dba3_0
[conda] pytorch                   1.0.1           cuda90py36h8b0c50b_0
[conda] tensorflow                1.13.1          mkl_py36h27d456a_0
[conda] tensorflow-base           1.13.1          mkl_py36h7ce6ba3_0
[conda] torchvision               0.2.1                    py36_0

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by vincentqb** [comment]
I confirm that I get:
```
>>> import torch
>>> divide_tensor = torch.Tensor([1.0, 0.0, 1.0])
>>> weight = torch.ones(3)
>>> weight.requires_grad = True
>>> tmp = weight / divide_tensor
>>> loss = tmp[0]
>>> loss.backward()
>>> print(weight.grad)
tensor([1., nan, 0.])
```

**Comment by ngimel** [comment]
This is a duplicate of #15506 and #15531

**Comment by albanD** [comment]
Similar to #15506 this cannot be solved.
The division operation's backward gets a gradient of 0 from the indexing operation for the part of the Tensor where you divide by 0. So the gradient it propagates is going to be `nan`.
You can change your code to do the indexing before the division to solve this:
```python
import torch
divide_tensor = torch.Tensor([1.0, 0.0, 1.0])
weight = torch.ones(3)
weight.requires_grad = True
tmp = weight[0] / divide_tensor[0]
loss.backward()
print(weight.grad)
```
